openml/extensions/sklearn/extension.py

```diff
@@ -1646,10 +1674,17 @@ def _prediction_to_probabilities(y: np.ndarray, classes: List[Any]) -> np.ndarra
                     # then we need to add a column full of zeros into the probabilities
                     # for class 3 because the rest of the library expects that the
                     # probabilities are ordered the same way as the classes are ordered).
-                    proba_y_new = np.zeros((proba_y.shape[0], len(task.class_labels)))
+
+                    # DataFrame allows more accurate mapping of classes as column names
+                    proba_y_new = pd.DataFrame(
+                        0,
+                        index=np.arange(proba_y.shape[0]),
+                        columns=task.class_labels,
+                        dtype=np.float32,
+                    )
                     for idx, model_class in enumerate(model_classes):
-                        proba_y_new[:, model_class] = proba_y[:, idx]
-                    proba_y = proba_y_new
+                        proba_y_new.loc[:, model_class] = proba_y[:, idx]
```

Comment:
Would it perhaps not be clear to:
 - keep the result of `prediction_to_probability` as a dataframe.
 - convert proba_y to a dataframe if `perdict_proba` was called:
    ```python
    proba_y = pd.DataFrame(proba_y, columns=model_classes)
    ```
 - then add missing columns
 - then reorder the dataframe:
   ```python
   proba_y = proba_y[task.class_labels]
   ```

Advantages:
 - Don't produce a whole copy in memory.
 - Keep indexing by class label only, never integers.
 - Can make (and keep) `_prediction_to_proba` store the probabilities as e.g. `uint8` for much smaller footprint (though conversion might be required later anyway, so I don't know if it is a win).

Reply:
That's a good spot. Thanks.
I made some changes to not have the extra copy and retain it as dataframe whenever applicable, without a conversion back and forth.

However, right at the end, needed to convert the probability array to numpy since [_run_task_get_arffcontent](https://github.com/openml/openml-python/blob/15743ee9f1e1cd5bfc772b7d67aec8d33724740d/openml/runs/functions.py#L504) appears to require a numpy as the probability matrix.

Reply:
> However, right at the end, needed to convert the probability array to numpy since _run_task_get_arffcontent appears to require a numpy as the probability matrix.

You could also change that function to accept a dataframe - it appears easier and safer to work with.

Reply:
Hey, I just checked but I think you did not yet update the return value type annotation for the functions `openml.extensions.sklearn.extension._prediction_to_probabilities` and `openml.extensions.sklearn.extension._run_model_on_fold`.

Reply:
Updated it, thanks for pointing it out.
However, not sure how to edit the docstrings since they don't appear to bear semblance to what the function is actually returning.

Reply:
That appears to be a wrong since two years, you can just update the docstring.

Reply:
Thanks for the update. I think Pieter's original suggestion is not yet addressed - do you think you could have a look whether that's still possible?

Reply:
is this resolved now then?